modules/videoio/src/cap_msmf.cpp

```diff
@@ -1155,41 +1424,66 @@ bool CvCapture_MSMF::retrieveFrame(int, cv::OutputArray frame)
-                    break;
-                case CV_CAP_MODE_BGR:
-                    if (captureMode == MODE_HW)
-                        cv::cvtColor(cv::Mat(captureFormat.height, captureFormat.width, CV_8UC4, ptr, pitch), frame, cv::COLOR_BGRA2BGR);
-                    else
-                        cv::Mat(captureFormat.height, captureFormat.width, CV_8UC3, ptr, pitch).copyTo(frame);
-                    break;
-                case CV_CAP_MODE_RGB:
-                    if (captureMode == MODE_HW)
-                        cv::cvtColor(cv::Mat(captureFormat.height, captureFormat.width, CV_8UC4, ptr, pitch), frame, cv::COLOR_BGRA2BGR);
-                    else
-                        cv::cvtColor(cv::Mat(captureFormat.height, captureFormat.width, CV_8UC3, ptr, pitch), frame, cv::COLOR_BGR2RGB);
-                    break;
-                case CV_CAP_MODE_GRAY:
-                    cv::Mat(captureFormat.height, captureFormat.width, CV_8UC1, ptr, pitch).copyTo(frame);
-                    break;
-                default:
-                    frame.release();
-                    break;
+                    switch (outputFormat)
+                    {
+                    case CV_CAP_MODE_YUYV:
+                        cv::Mat(captureVideoFormat.height, captureVideoFormat.width, CV_8UC2, ptr, pitch).copyTo(frame);
+                        break;
+                    case CV_CAP_MODE_BGR:
+                        if (captureMode == MODE_HW)
+                            cv::cvtColor(cv::Mat(captureVideoFormat.height, captureVideoFormat.width, CV_8UC4, ptr, pitch), frame, cv::COLOR_BGRA2BGR);
+                        else
+                            cv::Mat(captureVideoFormat.height, captureVideoFormat.width, CV_8UC3, ptr, pitch).copyTo(frame);
+                        break;
+                    case CV_CAP_MODE_RGB:
+                        if (captureMode == MODE_HW)
+                            cv::cvtColor(cv::Mat(captureVideoFormat.height, captureVideoFormat.width, CV_8UC4, ptr, pitch), frame, cv::COLOR_BGRA2BGR);
+                        else
+                            cv::cvtColor(cv::Mat(captureVideoFormat.height, captureVideoFormat.width, CV_8UC3, ptr, pitch), frame, cv::COLOR_BGR2RGB);
+                        break;
+                    case CV_CAP_MODE_GRAY:
+                        cv::Mat(captureVideoFormat.height, captureVideoFormat.width, CV_8UC1, ptr, pitch).copyTo(frame);
+                        break;
+                    default:
+                        frame.release();
+                        break;
+                    }
                 }
+                else
+                    frame.release();
             }
             else
-                frame.release();
+            {
+                cv::Mat(1, cursize, CV_8UC1, ptr, pitch).copyTo(frame);
+            }
         }
         else
         {
-            cv::Mat(1, cursize, CV_8UC1, ptr, pitch).copyTo(frame);
+            //switch(index)
```

Comment:
There is `cv::extractChannel()` call in OpenCV which can help to extract channel from interleaved data.